File: app/src/main/java/com/plantezeapp/HabitSuggestions/HabitMenu.java
package com.plantezeapp.HabitSuggestions;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Spinner;

import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.plantezeapp.R;

import java.util.List;


public class HabitMenu extends Fragment {

    public HabitMenu() {
        super(R.layout.fragment_habit_menu); // Pass the layout resource to the Fragment constructor
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        FloatingActionButton addHabits = view.findViewById(R.id.addHabit);

        //Navigate from HabitMenu to HabitCreation
        addHabits.setOnClickListener(v -> {
            NavController navController = Navigation.findNavController(view);
            navController.navigate(R.id.action_habitMenu_to_habitCreation);
        });

    }
}

File: app/src/main/java/com/plantezeapp/HabitSuggestions/Activities/HabitMenu.kt
package com.plantezeapp.HabitSuggestions.Activities

import android.content.Intent
import android.os.Bundle
import android.widget.SearchView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.plantezeapp.HabitSuggestions.Habit
import com.plantezeapp.HabitSuggestions.HabitAdapter
import com.plantezeapp.HabitSuggestions.HabitDao
import com.plantezeapp.HabitSuggestions.HabitDatabase
import com.plantezeapp.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import com.plantezeapp.Track.Tracker
import java.util.*

class HabitMenu : AppCompatActivity() {

    private lateinit var habitAdapter: HabitAdapter
    private lateinit var habitDao: HabitDao

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_habit_menu)

        val addHabitButton: FloatingActionButton = findViewById(R.id.addHabit)
        val recyclerView: RecyclerView = findViewById(R.id.rv_habits)
        val searchView: SearchView = findViewById(R.id.searchView)
        val goBackButton: FloatingActionButton = findViewById(R.id.goBack)

        // Initialize DAO
        val database = HabitDatabase.getDatabase(this)
        habitDao = database.habitDao()

        // Initialize HabitAdapter
        habitAdapter = HabitAdapter(mutableListOf(),
            { habitToDelete ->
                lifecycleScope.launch(Dispatchers.IO) {
                    habitDao.deleteHabit(habitToDelete)
                }
            },
            { habitToComplete ->
                onCompletedHabitClick(habitToComplete)
            }
        )

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = habitAdapter

        // Observe Habit List
        habitDao.getAllHabits().observe(this, { habits ->
            habits?.let {
                habitAdapter.updateData(it)
            }
        })

        // Search functionality
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                query?.let { filterHabits(it) }
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                if (newText.isNullOrEmpty()) {
                    habitDao.getAllHabits().observe(this@HabitMenu, { habits ->
                        habits?.let {
                            habitAdapter.updateData(it)
                        }
                    })
                } else {
                    filterHabits(newText)
                }
                return true
            }
        })

        // Navigate to HabitCreation
        addHabitButton.setOnClickListener {
            val intent = Intent(this, HabitCreation::class.java)
            startActivity(intent)
        }

        // Navigate to Tracker
        goBackButton.setOnClickListener {
            val intent = Intent(this, Tracker::class.java)
            startActivity(intent)
        }
    }

    private fun filterHabits(query: String) {
        habitDao.searchHabits("%$query%").observe(this, { habits ->
            habits?.let {
                habitAdapter.updateData(it)
            }
        })
    }

    private fun onCompletedHabitClick(habit: Habit) {
        // Increment the completion count
        val updatedHabit = habit.copy(completionCount = habit.completionCount + 1)

        // Update the habit in the database
        lifecycleScope.launch(Dispatchers.IO) {
            habitDao.updateHabit(updatedHabit)
        }

        // Notify the adapter to update the specific item
        habitAdapter.notifyDataSetChanged()

        // Show a Toast message
        Toast.makeText(this, "Habit Completed!", Toast.LENGTH_SHORT).show()
    }

}

